### PR TITLE
Changes to enable usage of prettyprinter backends

### PR DESCRIPTION
**Pull request by georgefst**
Using `prettyprinter-ansi-terminal`'s `AnsiStyle` to represent colours was not ideal as it is intentionally opaque, and so it's difficult to convert to another format.

This PR introduces a new `Style` type corresponding to the subset of `AnsiStyle` which is used in `pretty-simple`.

There is some inconsistency in whether we refer to things as "style"s or "colour"s. This `Style` type allows customising whether the text is bold, so it does represent more than just colour. But changing all instances of "color" to "style" would seem an unnecessary break of backwards compatibility. This isn't really a new issue, but this change has made it more obvious.

### CODE REVIEW

File: src/Text/Pretty/Simple/Internal/OutputPrinter.hs [hunk 1]
```diff
@@ -188,9 +187,9 @@ hCheckTTY h options = liftIO $ conv <$> tty
 -- | Parse a string, and generate an intermediate representation,
 -- suitable for passing to any /prettyprinter/ backend.
 -- Used by 'Simple.pString' etc.
-layoutString :: OutputOptions -> String -> SimpleDocStream AnsiStyle
+layoutString :: OutputOptions -> String -> SimpleDocStream Style
 layoutString opts =
```
Comment (georgefst) [inline src/Text/Pretty/Simple/Internal/OutputPrinter.hs:191]: Perhaps we should actually document this functionality properly, and expose this function from a non-Internal module?

File: src/Text/Pretty/Simple/Internal/Color.hs [hunk 2]
```diff
@@ -25,262 +27,110 @@ module Text.Pretty.Simple.Internal.Color
...
-colorDullCyan = colorDull Cyan
-
-colorDullGreen :: AnsiStyle
-colorDullGreen = colorDull Green
-
-colorDullMagenta :: AnsiStyle
-colorDullMagenta = colorDull Magenta
-
-colorDullRed :: AnsiStyle
-colorDullRed = colorDull Red
```
Comment (georgefst) [inline src/Text/Pretty/Simple/Internal/Color.hs:268]: I may have gone overboard in removing things from this module. But with this more user-friendly colour type (previously it was just `Builder`), it seems a bit pointless to export all these combinations.

File: src/Text/Pretty/Simple.hs [hunk 3]
```diff
@@ -116,11 +122,16 @@ import Control.Applicative
 
 import Control.Monad.IO.Class (MonadIO, liftIO)
 import Data.Text.Lazy (Text)
-import Prettyprinter.Render.Terminal (renderLazy, renderIO)
+import Prettyprinter (SimpleDocStream)
+import Prettyprinter.Render.Terminal
+      (Color (..), Intensity(Vivid,Dull), AnsiStyle,
+       renderLazy, renderIO)
```
Comment (georgefst) [inline src/Text/Pretty/Simple.hs:128]: I haven't quite been able to work out the formatting rules for imports. Do you use a particular formatter?

**Comment by georgefst** [comment]
There are a few missing Haddocks. I'll add those once everything else is approved.

**Comment by cdepillabout** [comment]
@georgefst You've been doing a bunch of work on `pretty-simple` lately (including various dependencies of `pretty-simple`).

Would you be interested in become a maintainer of `pretty-simple`?  That way, you could review PRs, as well as merge your own PRs and make releases without having to wait for me.

I've been pretty busy at work lately, and I haven't had time to take a good look at your last couple PRs.

**Comment by georgefst** [comment]
> Would you be interested in become a maintainer of `pretty-simple`?

That would be great. If you're happy not to review the current PRs in detail, then I'll clean them up and merge. I'd still feel more comfortable checking in with you for at least a high-level overview of changes before publishing a release (and I really don't mind waiting a while for that if necessary).

**Comment by cdepillabout** [comment]
@georgefst I sent you an invitation to be a collaborator for this repo.  I think you should get an email from github about it.

Also, if you send me your hackage username, I'll add you as a maintainer on Hackage.

> I'd still feel more comfortable checking in with you for at least a high-level overview of changes before publishing a release 

I'd definitely be fine with checking for a high-level overview of changes, although you can feel free to go ahead and merge stuff yourself, as well as review/merge PRs from other users.

For making a release to Hackage, I have a checklist: https://functor.tokyo/blog/2018-07-16-release-haskell-packages-to-hackage

Most of these should be pretty obvious, but the two important points are to make sure to update the `CHANGELOG.md` before doing a release, and then to make sure to create a tag and push it to GitHub on the commit that you uploaded to Hackage.

------------------------------------------

As far as the long-term vision for `pretty-simple`, the two points that are the most important to me are the following:

- `pretty-simple` should be simple to use.  Most users should just be able to import `pPrint` and use it.
- `pretty-simple` should try to hard prettify and print anything you throw at it (even if it is not valid Haskell code).  The output doesn't have to _perfect_ but it should at least be better than Haskell's normal `print` output.

File: src/Text/Pretty/Simple.hs (hunk 3 above)
Comment (cdepillabout) [reply src/Text/Pretty/Simple.hs:128]: `hindent` used to format imports like this (but now it has been changed to format import lists differently).

However, if you want to change this (or even run a formatter over all the code), please feel free.

**Comment by cdepillabout** [comment]
I took a quick look at this, and I'm happy with it if you want to merge it in!

**Comment by georgefst** [comment]
@cdepillabout I'm [GeorgeThomas](https://hackage.haskell.org/user/GeorgeThomas) on Hackage.

**Comment by cdepillabout** [comment]
Thanks, I've added you on Hackage: https://hackage.haskell.org/package/pretty-simple/maintainers/

**Pull request merged**